app/jobs/back_in_stock_notification_email_job.rb

```diff
@@ -10,8 +10,11 @@ def perform(stock_location_params: {})
   private
 
   def send_back_in_stock_notifications
-    in_stock_user_notifications.each do |email, user_bisns|
-      Spree::BackInStockNotificationMailer.notification(user_bisns).deliver_now
+    in_stock_user_notifications.each_with_index do |in_stock_user_notification, index|
+      email, user_bisns = in_stock_user_notification
+      delay = (Spree::BackInStockNotification::EMAIL_DELAY_FACTOR * (index+1)).seconds
+
+      Spree::BackInStockNotificationMailer.notification(user_bisns).deliver_later(wait: delay)
```

Comment:
@jarednorman it occurred to me that by using a deliver-later call here could result in the email failing to send and this, this job not know about and incorrectly mark the email as sent. Any suggestions how to deal with this? Or do you think my concerns about causing problems by sending the emails all at once where unfounded?

Reply:
You could lock the notification record in the job to prevent it from being accessed concurrently and then record the status on the notification.

Reply:
@jarednorman can you give an example? I'm not sure how a lock would work here considering several records can be included in an email. Did you mean updating the email status in the mailer job as per committed example? If so, I'm not sure how best to test this.

Reply:
Yeah, I'm not sure I've looked closely enough at how this all works. Usually I try not to do "work" (anything not related to composing the email) in a worker to make the mailers as "dumb" as possible. Perhaps you could do something like have this job query for users that have notifications that need sending and queue jobs to send for each user.

The jobs that send to each user could lock on the user row to avoid concurrently processing notifications for that user and could only query for notifications that haven't been processed, aborting if none are found. This way there's no problem with the job getting queued 100 times for one user, as they'll still only ever get the notification once:

```ruby
class UserBackInStockNotificationJob < ApplicationJob
  def perform(user_id)
    ActiveRecord::Base.transaction do
      user = Spree::User.lock.find(user_id)
      notifications = user.notifications.ready_to_send
      return unless notifications.any?
      
      # send email
      
      # record that you sent the email
    end
  end
end
```

Reply:
@jarednorman that looks good. Unfortunately because we don't use guest user sessions most of the notifications are not associated with a user - just an email address. Perhaps it would make sense to create a "Back In Stock" user, which would be helpful for storing stock location and language options etc.

As that will be a bit of reworking of the existing implementation I am inclined to add another job to handle the sending of the email and running it with `deliver_later` on live and `deliver_now` in test.